core/src/main/scala/io/projectglow/gff/GffDataSource.scala

```diff
@@ -0,0 +1,346 @@
+   *       attribute fields will have StringType.
+   * Ordering: gffBaseSchema fields come first, followed by official attributes fields as ordered
+   *           in [[io.projectglow.common.FeatureSchemas.gffOfficialAttributeFields]], followed by
+   *           unofficial attribute fields in case-insensitive alphabetical order.
+   *
+   * @param sqlContext
+   * @param path: path to gff file or glob
+   * @return inferred schema
+   */
+  def inferSchema(sqlContext: SQLContext, path: String): StructType = {
+    val spark = sqlContext.sparkSession
+
+    val originalColumnPruning: Boolean =
+      spark.conf.get(columnPruningConf).toBoolean
+
+    // csv column pruning is turned off temporarily for the malformed rows (such as Fasta rows) to
+    // get filtered by DROPMALFORMED mode option of csv reader no matter which columns are selected
+    // in the query. See https://issues.apache.org/jira/browse/SPARK-28058
+    if (originalColumnPruning) {
+      spark.conf.set(columnPruningConf, false)
+    }
+
+    val csvDf = spark
+      .read
+      .options(csvReadOptions)
+      .schema(gffBaseSchema)
+      .csv(path)
+
+    val attributeTags = addAttributesMapColumn(csvDf)
+      .withColumn(
+        attributesMapColumnName,
+        explode(
+          map_keys(
+            col(attributesMapColumnName)
+          )
+        )
+      )
+      .agg(
+        collect_set(
+          attributesMapColumnName
+        )
+      )
+      .collect()(0)
+      .getAs[Seq[String]](0)
+      .filter(!_.isEmpty)
+
+    spark.conf.set(columnPruningConf, originalColumnPruning)
+
+    val attributeFields = attributeTags
+      .foldLeft(Seq[StructField]()) { (s, t) =>
+        val officialIdx = officialFieldNames.indexOf(normalizeString(t))
+        if (officialIdx > -1) {
+          s :+ StructField(t, gffOfficialAttributeFields(officialIdx).dataType)
+        } else {
+          s :+ StructField(t, StringType)
+        }
+      }
+      .sortBy(_.name)(FieldNameOrdering)
+
+    StructType(
+      gffBaseSchema.fields.dropRight(1) ++ attributeFields
+    )
+  }
+
+  def addAttributesMapColumn(df: DataFrame): DataFrame = {
+    df.withColumn(
+      attributesMapColumnName,
+      expr(
+        s"""str_to_map(
+           |       ${attributesField.name},
+           |       "$ATTRIBUTES_DELIMITER",
+           |       "$GFF3_TAG_VALUE_DELIMITER"
+           |   )""".stripMargin
+      )
+    )
+  }
+
+  def normalizeAttributesMapKeys(df: DataFrame): DataFrame = {
+    df.withColumn(
+      attributesMapColumnName,
+      map_from_arrays(
+        expr(
+          s"""transform(
+            |       map_keys($attributesMapColumnName),
+            |       k -> regexp_replace(lower(k), "_", "")
+            |   )""".stripMargin
+        ),
+        map_values(col(attributesMapColumnName))
+      )
+    )
+  }
+
+  def filterFastaLines(df: DataFrame): DataFrame = {
+    df.where(
```

Comment:
This is kind of a confusing way to write this filter. `fold` is specifically called out in our style guide as a hard-to-understand pattern: https://github.com/databricks/scala-style-guide. [Coalesce](https://spark.apache.org/docs/2.4.0/api/scala/index.html#org.apache.spark.sql.functions$@coalesce(e:org.apache.spark.sql.Column*):org.apache.spark.sql.Column) might be a good alternative here.

Reply:
Used Coalesce